packages/plugin-ext/src/main/node/plugin-deployer-impl.ts

```diff
@@ -296,40 +296,38 @@ export class PluginDeployerImpl implements PluginDeployer {
      * If there are some single files, try to see if we can work on these files (like unpacking it, etc)
      */
     public async applyFileHandlers(pluginDeployerEntries: PluginDeployerEntry[]): Promise<any> {
-        const waitPromises: Array<Promise<any>> = [];
-
-        pluginDeployerEntries.filter(pluginDeployerEntry => pluginDeployerEntry.isResolved()).map(pluginDeployerEntry => {
-            this.pluginDeployerFileHandlers.map(pluginFileHandler => {
+        const waitPromises = pluginDeployerEntries.filter(pluginDeployerEntry => pluginDeployerEntry.isResolved()).map(pluginDeployerEntry =>
+            this.pluginDeployerFileHandlers.reduce((acc, pluginFileHandler) => {
                 const proxyPluginDeployerEntry = new ProxyPluginDeployerEntry(pluginFileHandler, (pluginDeployerEntry) as PluginDeployerEntryImpl);
-                if (pluginFileHandler.accept(proxyPluginDeployerEntry)) {
-                    const pluginDeployerFileHandlerContext: PluginDeployerFileHandlerContext = new PluginDeployerFileHandlerContextImpl(proxyPluginDeployerEntry);
-                    const promise: Promise<void> = pluginFileHandler.handle(pluginDeployerFileHandlerContext);
-                    waitPromises.push(promise);
-                }
-            });
-
-        });
-        return Promise.all(waitPromises);
+                acc.push(async () => {
+                    if (await pluginFileHandler.accept(proxyPluginDeployerEntry)) {
+                        const pluginDeployerFileHandlerContext = new PluginDeployerFileHandlerContextImpl(proxyPluginDeployerEntry);
+                        await pluginFileHandler.handle(pluginDeployerFileHandlerContext);
+                    }
+                });
+                return acc;
+            }, [] as (() => Promise<void>)[])
+        ).reduce((left, right) => left.concat(right), []);
+        return Promise.all(waitPromises.map(promise => promise()));
```

Comment:
Suggestion: This looks overly complicated. Why not use `flatMap`? Same for the `applyDirectoryFileHandlers` method below.

```ts
const waitPromises = pluginDeployerEntries.filter(pluginDeployerEntry => pluginDeployerEntry.isResolved()).flatMap(pluginDeployerEntry =>
    this.pluginDeployerFileHandlers.map(async pluginFileHandler => {
        const proxyPluginDeployerEntry = new ProxyPluginDeployerEntry(pluginFileHandler, (pluginDeployerEntry) as PluginDeployerEntryImpl);
        if (await pluginFileHandler.accept(proxyPluginDeployerEntry)) {
            const pluginDeployerFileHandlerContext = new PluginDeployerFileHandlerContextImpl(proxyPluginDeployerEntry);
           await pluginFileHandler.handle(pluginDeployerFileHandlerContext);
       }
    })
);
return Promise.all(waitPromises);
```

Reply:
https://github.com/eclipse-theia/theia/pull/12798/commits/e78a76148e43b02473c899f8090b2514365f2e22